Fix extra heap allocation in Source constructor

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151850
* #151849
* #151810
* #151807
* #151806
* #151805
* #151804
* #151803
* #151802
* #151801
* __->__ #151800
* #151682

This was a sneaky one: the StringCordView default constructor allocates.

Differential Revision: [D73129448](https://our.internmc.facebook.com/intern/diff/D73129448/)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

swolchok committed:
Fix extra heap allocation in Source constructor

This was a sneaky one: the StringCordView default constructor allocates.

Differential Revision: [D73129448](https://our.internmc.facebook.com/intern/diff/D73129448/)

[ghstack-poisoned]

swolchok committed:
Update on "Fix extra heap allocation in Source constructor"


This was a sneaky one: the StringCordView default constructor allocates.

Differential Revision: [D73129448](https://our.internmc.facebook.com/intern/diff/D73129448/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

malfet commented on torch/csrc/jit/frontend/source_range.h:
```diff
@@ -287,6 +280,18 @@ struct TORCH_API Source {
+      CopiesString copies_str,
+      std::string_view text_view) {
+    if (copies_str == COPIES_STRING) {
+      std::shared_ptr<std::string> allocated_str =
```
Nit
```suggestion
      auto allocated_str =
```

swolchok committed:
Update on "Fix extra heap allocation in Source constructor"


This was a sneaky one: the StringCordView default constructor allocates.

Differential Revision: [D73129448](https://our.internmc.facebook.com/intern/diff/D73129448/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

cyyever commented:
@pytorchmergebot merge -i

Skylion007 commented on torch/csrc/jit/frontend/source_range.h:
```diff
@@ -287,6 +280,18 @@ struct TORCH_API Source {
+    if (copies_str == COPIES_STRING) {
+      auto allocated_str =
+          std::make_shared<std::string>(text_view.data(), text_view.size());
+      return StringCordView({*allocated_str}, {allocated_str});
```
Allocated string shouldn't be moved in the second arg?

swolchok replied:
whoops, just saw this. PR is already landed internally and there appears to be some propagation delay, so I can't fix it on this PR.

just adding a `std::move` wouldn't work because order of subexpression evaluation is unspecified, so `*allocated_str` might break. you can still avoid the refcount bump, but I'm probably not going to get to it unless a profiler tells me to because this path is also doing three heap allocations (the StringCordView arguments are vectors) so it's a bit of a lost cause for efficiency anyway. (the TorchLibraryInit path isn't COPIES_STRING)

swolchok committed:
rebase and minor fixes as requested on "Fix extra heap allocation in Source constructor"


This was a sneaky one: the StringCordView default constructor allocates.

Differential Revision: [D73129448](https://our.internmc.facebook.com/intern/diff/D73129448/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]
